janusgraph-core/src/main/java/org/janusgraph/graphdb/configuration/GraphDatabaseConfiguration.java

```diff
@@ -1286,15 +1301,46 @@ public GraphDatabaseConfiguration(ReadConfiguration localConfig) {
+                else
+                    allowUpgrade = ALLOW_UPGRADE.getDefaultValue();
+
+                if (allowUpgrade) {
+                    // If the graph doesn't have a storage version set it and update version
+                    if (!globalWrite.has(INITIAL_STORAGE_VERSION)) {
+                        globalWrite.set(INITIAL_JANUSGRAPH_VERSION, JanusGraphConstants.VERSION);
+                        globalWrite.set(INITIAL_STORAGE_VERSION, JanusGraphConstants.STORAGE_VERSION);
+                        log.info("graph.storage-version has been upgraded from 1 to {} and graph.janusgraph-version has been upgraded from {} to {} on graph {}",
+                                 JanusGraphConstants.STORAGE_VERSION, globalWrite.get(INITIAL_JANUSGRAPH_VERSION), JanusGraphConstants.VERSION, getGraphName());
+                    // If the graph has a storage version, but it's lower than the client or server opening the graph upgrade the version and storage version
+                    } else if (Integer.parseInt(globalWrite.get(INITIAL_STORAGE_VERSION)) < Integer.parseInt(JanusGraphConstants.STORAGE_VERSION)) {
+                        globalWrite.set(INITIAL_JANUSGRAPH_VERSION, JanusGraphConstants.VERSION);
+                        globalWrite.set(INITIAL_STORAGE_VERSION, JanusGraphConstants.STORAGE_VERSION);
+                        log.info("graph.storage-version has been upgraded from {} to {} and graph.janusgraph-version has been upgraded from {} to {} on graph {}",
+                                 globalWrite.get(INITIAL_STORAGE_VERSION), JanusGraphConstants.STORAGE_VERSION, globalWrite.get(INITIAL_JANUSGRAPH_VERSION), JanusGraphConstants.VERSION, getGraphName());
+                    // If the storage version of the client or server opening the graph is lower than the graph's storage version throw an exception
+                    } else if (Integer.parseInt(globalWrite.get(INITIAL_STORAGE_VERSION)) > Integer.parseInt(JanusGraphConstants.STORAGE_VERSION)) {
+                        throw new JanusGraphException(String.format(BACKLEVEL_STORAGE_VERSION_EXCEPTION, globalWrite.get(INITIAL_STORAGE_VERSION), JanusGraphConstants.STORAGE_VERSION, getGraphName()));
+                    } else {
+                        log.warn("Warning graph.allow-upgrade is currently set to true on graph {}. Please remove graph.allow-upgrade from your properties file.", getGraphName());
+                    }
+
+                } else {
+                         if (!globalWrite.has(INITIAL_STORAGE_VERSION) || !globalWrite.get(INITIAL_STORAGE_VERSION).equals(JanusGraphConstants.STORAGE_VERSION)) {
+                            String storageVersion = (globalWrite.has(INITIAL_STORAGE_VERSION)) ? globalWrite.get(INITIAL_STORAGE_VERSION) : "1";
+                            throw new JanusGraphException(String.format(INCOMPATIBLE_STORAGE_VERSION_EXCEPTION, storageVersion, JanusGraphConstants.STORAGE_VERSION, getGraphName()));
```

Comment:
`getGraphName()` can return `null`. It would be a nicer error message if it used an empty string rather than `null`.